Step2 지하철 구간 제거 기능 개선

yunjiji opened the pull request:
지하철 구간 제거 기능 개선하였습니다.

기존의 addSection에서 썼던 함수를 리팩토링하여 쓰긴 했는데
리팩토링을 했을 때 만족스럽지가 않네요

Commit:
2단계 : 지하철 구간 제거 기능 개선

Commit:
2단계 : 지하철 구간 제거 기능 개선

LineServiceTest deleteSection 추가

dia312 commented on src/main/java/nextstep/subway/domain/Line.java:
```diff
@@ -40,32 +43,56 @@ public void addSection(Section section) {
     }
 
+    public void deleteSection(Station station) {
+        if(sections.size() == 1) {
+            throw new SectionDeleteException("구간이 하나인 경우 구간을 삭제할 수 없습니다.");
+        }
+        if(!isExistedStation(station)) {
+            throw new SectionDeleteException("노선에 등록되지 않은 역은 제거할 수 없습니다.");
+        }
```
List<Sections>을 일급 컬렉션(Sections)로 만들어 위 유효성 검사를 sections에 위임하면 어떨까요?

dia312 commented on src/main/java/nextstep/subway/domain/Line.java:
```diff
@@ -40,32 +43,56 @@ public void addSection(Section section) {
+            throw new SectionDeleteException("노선에 등록되지 않은 역은 제거할 수 없습니다.");
+        }
+
+        Optional<Section> upSection = findSectionByUpStation(station);
+        Optional<Section> downSection = findSectionByDownStation(station);
```
Optional 객체로 return 받기보다 find 메서드 내에서 Section을 못찾을 경우 예외를 던져보면 어떨까요?

dia312 commented on src/main/java/nextstep/subway/domain/Line.java:
```diff
@@ -40,32 +43,56 @@ public void addSection(Section section) {
+        Optional<Section> upSection = findSectionByUpStation(station);
+        Optional<Section> downSection = findSectionByDownStation(station);
+
+        if(upSection.isEmpty() && downSection.isPresent()) {
+            sections.remove(downSection.get());
+
+        } else if(upSection.isPresent() && downSection.isEmpty()) {
+            sections.remove(upSection.get());
+
+        } else if(upSection.isPresent() && downSection.isPresent()) {
+            upSection.get().modifyUp(downSection.get().getUpStation());
+            upSection.get().modifyDistanceForRemove(downSection.get());
+            sections.remove(downSection.get());
+        }
```
> [객체지향 생활체조 원칙] 규칙 2. else 예약어를 쓰지 않는다

위 원칙에 따라 리팩토링해볼 수 잇을까요?

yunjiji replied:
이 부분을 수정하려는데 이 케이스를 어떻게 분리해야할 지 모르겠어요
비어있을 경우 Exception을 내보내려니 지우려는 station이 상행역 또는 하행역일 때 에러가 나서 Optional로 변경을 한건데
갯수로 파악을 해야하는걸까요? 
상행역 하행역인 경우와 가운데 껴있는 경우 케이스로 분리는 가능할 것 같은데 .. 뭔가 깔끔하지가 않네요 :(

dia312 commented on src/test/java/nextstep/subway/acceptance/SectionAcceptanceTest.java:
```diff
@@ -117,6 +117,44 @@ void removeLineSection() {
         assertThat(response.jsonPath().getList("stations.id", Long.class)).containsExactly(강남역, 양재역);
     }
 
+    /**
```
실패하는 예외케이스까지 테스트해보셨네요 👍

dia312 replied:
추가된 구간 제거 기능이 성공하는 테스트도 작성해볼 수 있을까요?

dia312 reviewed (changes requested):
안녕하세요 지민님!
2단계 미션 수행 잘해주셨네요 👍 
몇가지 코멘트 남긴 부분 있는데 확인해주시고 리뷰요청 부탁드릴게요!

dia312 commented on src/main/java/nextstep/subway/domain/Section.java:
```diff
@@ -35,12 +35,15 @@ public Section(Line line, Station upStation, Station downStation, int distance)
     }
 
-    public void modifyDistance(Section section){
+    public void modifyDistanceForAdd(Section section){
         if(this.distance <= section.getDistance()){
             throw new SectionDistanceException("등록할 구간의 길이가 기존 역 사이 길이보다 크거나 같습니다.");
         }
         this.distance -= section.getDistance();
     }
+    public void modifyDistanceForRemove(Section section){
+        this.distance += section.getDistance();
+    }
```
Section을 넘겨주지 않고 section의 거리만 넘겨줘도 될 것 같은데 어떻게 생각하시나요?

Commit:
2단계 : 지하철 구간 제거 기능 개선

코멘트 반영
- 가운데 구간 제거 테스트 추가
- Sections 일급컬렉션 추가
- Sections 에 함수 일임
- Section 대신 distance로 넘김

dia312 commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -0,0 +1,132 @@
+        boolean isDown = isExistedDownStation(station);
+
+        // up
+        if(isUp && !isDown) {
+            sections.remove(findSectionByDownStation(station));
+        }
+        //down
+        if(!isUp && isDown){
+            sections.remove(findSectionByUpStation(station));
+        }
+        // middle
+        if(isUp && isDown) {
+            Section up = findSectionByUpStation(station);
+            Section down = findSectionByDownStation(station);
+            up.modifyUp(down.getUpStation());
+            up.modifyDistanceForRemove(down.getDistance());
+            sections.remove(down);
+        }
```
피드백 반영 👍  
삭제 로직이 복잡해서 지민님이 리팩토링한 코드만으로도 충분히 가독성이 좋아진 것 같아요!
또한 findSectionByUpStation이나 findSectionByDownStation 에서 상행역과 하행역을 찾지 못하는 경우, 
이전 코드처럼 예외를 던져주지 않는 것보다 정확한 예외메세지("구간이 존재하지 않는다")를 던져 주는 것이 더 올바른 처리 방법이라고 생각해요 :)

dia312 commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -0,0 +1,132 @@
+        return sections.stream()
+                .filter(section -> section.isUp(station))
+                .findAny()
+                .orElseThrow(() -> new SectionNotFoundException("구간을 찾을 수 없습니다."));
+    }
+    public Section findSectionByDownStation(Station station) {
+        return sections.stream()
+                .filter(section -> section.isDown(station))
+                .findAny()
+                .orElseThrow(() -> new SectionNotFoundException("구간을 찾을 수 없습니다."));
+    }
+
+    public Section findSectionByUpAndDownStation(Station up, Station down) {
+        return sections.stream()
+                .filter(section -> section.isUp(up) && section.isDown(down))
+                .findAny()
+                .orElseThrow(() -> new SectionNotFoundException("구간을 찾을 수 없습니다."));
+    }
```
동일한 예외 메세지가 중복돼서 사용되네요! 
예외 메세지를 관리하는 Enum이나 Class를 활용해보면 어떨까요?

dia312 commented on src/test/java/nextstep/subway/acceptance/SectionAcceptanceTest.java:
```diff
@@ -108,6 +108,28 @@ void removeLineSection() {
+     * When 지하철 노선의 가운데 있는 역을 제거하면
+     * Then 노선에 구간이 제거된다
+     */
+    @DisplayName("지하철 노선에 구간을 제거")
+    @Test
+    void removeLineSectionMiddle() {
+        // given
+        Long 정자역 = 지하철역_생성_요청("정자역").jsonPath().getLong("id");
+        지하철_노선에_지하철_구간_생성_요청(신분당선, createSectionCreateParams(양재역, 정자역, 6));
```
```suggestion
    @DisplayName("지하철 노선에 가운데 역을 포함한 구간을 제거")
    @Test
    void removeLineSectionMiddle() {
        // given
        Long 정자역 = 지하철역_생성_요청("정자역").jsonPath().getLong("id");
        지하철_노선에_지하철_구간_생성_요청(신분당선, createSectionCreateParams(양재역, 정자역, 6));
```
👍

dia312 reviewed (approved):
안녕하세요 지민님!
피드백 반영 잘해주셨네요 ㅎㅎ 몇가지 코멘트 남긴 부분 확인해주시고 다음 단계 진행해주세요!
궁금한 부분이 있으면 언제든지 물어봐주세요 😄
